hub/core/storage/s3/s3_mapper.py

```diff
@@ -0,0 +1,173 @@
+        """
+        try:
+            path = posixpath.join(self.path, path)
+            content = bytearray(memoryview(content))
+            attrs = {
+                "Bucket": self.bucket,
+                "Body": content,
+                "Key": path,
+                "ContentType": ("application/octet-stream"),
+            }
+            self.client.put_object(**attrs)
+        except Exception as err:
+            raise S3SetError(err)
+
+    def __getitem__(self, path):
+        """
+        Gets the object present at the path.
+
+        Args:
+            path (str): the path relative to the root of the mapper.
+
+        Returns:
+            bytes: The bytes of the object present at the path.
+
+        Raises:
+            KeyError: If an object is not found at the path.
+            S3GetError: Any other error other than KeyError while retrieving the object.
+        """
+        try:
+            path = posixpath.join(self.path, path)
+            resp = self.client.get_object(
+                Bucket=self.bucket,
+                Key=path,
+            )
+            return resp["Body"].read()
+        except ClientError as err:
+            if err.response["Error"]["Code"] == "NoSuchKey":
+                raise KeyError(err)
+            raise
+        except Exception as err:
+            raise S3GetError(err)
+
+    def __delitem__(self, path):
+        """
+        Delete the object present at the path.
+
+        Args:
+            path (str): the path to the object relative to the root of the mapper.
+
+        Returns:
+            None
+
+        Raises:
+            S3DeletionError: Any S3 error encountered while deleting the object. Note: if the object is not found, s3 won't raise KeyError.
+        """
+        try:
+            path = posixpath.join(self.path, path)
+            self.client.delete_object(Bucket=self.bucket, Key=path)
+        except Exception as err:
+            raise S3DeletionError(err)
+
+    def _list_objects(self):
+        """
+        Helper function to list all the objects present at the root of the mapper.
+
+        Args:
+            None
+
+        Returns:
+            list: list of all the objects found at the root of the mapper.
+
+        Raises:
+            S3ListError: Any S3 error encountered while listing the objects.
+        """
+        try:
+            # TODO boto3 list_objects only returns first 1000 objects
+            items = self.client.list_objects_v2(Bucket=self.bucket, Prefix=self.path)
+            items = items["Contents"]
+            names = [item["Key"] for item in items]
+            # removing the prefix from the names
+            len_path = len(self.path.split("/"))
+            names = ["/".join(name.split("/")[len_path:]) for name in names]
+            return names
+        except Exception as err:
+            raise S3ListError(err)
+
+    def __len__(self):
+        """
+        Returns the number of files present inside the root of the mapper.
```

Comment:
Would be nice to have a note here that this is expensive, and makes a call every time. `len` is normally assumed to be cheap, but it isn't here.